databuilder/extractor/bigquery_usage_extractor.py

```diff
@@ -0,0 +1,103 @@
+        self.table_usage_counts = {}
+        self._count_usage()
+        self.iter = iter(self.table_usage_counts)
+
+    def _count_usage(self):
+        for entry in self._retrieve_records():
+            if not isinstance(entry, dict):
+                raise Exception("BigQueryUsageExtractor expects record of type dict!")
+
+            email = entry['protoPayload']['authenticationInfo']['principalEmail']
+            jobCompletedEvent = entry['protoPayload']['serviceData']['jobCompletedEvent']
+            refTables = jobCompletedEvent['job']['jobStatistics'].get('referencedTables', None)
+
+            if not refTables:
+                # Query results can be cached and if the source tables remain untouched,
+                # bigquery will return it from a 24 hour cache result instead.
+                continue
+
+            for refTable in refTables:
+                key = TableColumnUsageTuple(database='bigquery',
+                                            cluster=refTable['projectId'],
+                                            schema=refTable['datasetId'],
+                                            table=refTable['tableId'],
+                                            column='*',
+                                            email=email)
+                new_count = self.table_usage_counts.get(key, 0) + 1
+                self.table_usage_counts[key] = new_count
+
+    def _retrieve_records(self):
+        # type: () -> Optional[TableColumnUsage]
+        """
+        Extracts bigquery log data by looking at the principalEmail in the
+        authenticationInfo block and referencedTables in the jobStatistics.
+
+        :return: Provides a record or None if no more to extract
+        """
+        body = {
+            'resourceNames': [
+                'projects/{projectid}'.format(projectid=self.projectid)
+            ],
+            'pageSize': 300,
+            'filter': 'resource.type="bigquery_resource" AND '
+                      'protoPayload.methodName="jobservice.jobcompleted" AND '
+                      'timestamp >= "{timestamp}"'.format(timestamp=self.timestamp)
+        }
+        for page in self._page_over_results(body):
+            for entry in page['entries']:
+                yield(entry)
+
+    def extract(self):
+        try:
+            key = next(self.iter)
+            return key, self.table_usage_counts[key]
+        except StopIteration:
+            return None
+
+    def _page_over_results(self, body):
+        response = self.logging_service.entries().list(body=body).execute(num_retries=7)
+        while response:
+            yield response
+
+            if 'nextPageToken' in response:
+                body['pageToken'] = response['nextPageToken']
+                response = self.logging_service.entries().list(body=body).execute(num_retries=7)
+            else:
+                response = None
+
+    def get_scope(self):
+        # type: () -> str
+        return 'extractor.bigquery_usage'
```

Comment:
How about `extractor.bigquery_usage_aggregate` ?

Reply:
I have opted for bigquery_table_usage instead , because we're not actually aggregating from columns to table level, just selecting referenced tables from the API.

Reply:
I see. Actually the term `aggregate` is used in `TblColUsgAggExtractor` as it aggregates the count like aggregate function `SUM`.

(The first attempt was extractor to just pass usage info and mutate Neo4j when published, but the idea was dropped, thus introduced `TblColUsgAggExtractor` )